src/shared/creds-util.c

```diff
@@ -1652,3 +1654,91 @@ int ipc_decrypt_credential(const char *validate_name, usec_t validate_timestamp,
 
         return 0;
 }
+
+static int pick_up_credential_one(
+                int credential_dir_fd,
+                const char *credential_name,
+                const PickUpCredential *table_entry) {
+
+        _cleanup_free_ char *fn = NULL, *target_path = NULL;
+        _cleanup_close_ int target_dir_fd = -EBADF;
+        const char *e;
+        int r;
+
+        assert(credential_dir_fd >= 0);
+        assert(credential_name);
+        assert(table_entry);
+
+        e = startswith(credential_name, table_entry->credential_prefix);
+        if (!e)
+                return 0; /* unmatched */
+
+        fn = strjoin(e, table_entry->filename_suffix);
+        if (!fn)
+                return log_oom();
+
+        if (!filename_is_valid(fn))
+                return log_warning_errno(SYNTHETIC_ERRNO(EINVAL),
+                                         "Passed credential '%s' would result in invalid filename '%s', ignoring.",
+                                         credential_name, fn);
+
+        target_dir_fd = xopenat_full(AT_FDCWD, table_entry->target_dir, O_CLOEXEC | O_CREAT | O_DIRECTORY, /* xopen_flags = */ 0, 0755);
+        if (target_dir_fd < 0)
+                return log_warning_errno(target_dir_fd, "Failed to open %s: %m", table_entry->target_dir);
+
+        target_path = path_join(table_entry->target_dir, fn);
+        if (!target_path)
+                return log_oom();
+
+        r = copy_file_at(
+                        credential_dir_fd, credential_name,
+                        target_dir_fd, fn,
```

Comment:
If dir_fd is not used, `fn` needs to be replaced by `target_path`

Reply:
Yes, but I am a big fan of using dirfd. Why not to use it when we can?